qubes/ext/core_features.py

```diff
@@ -0,0 +1,62 @@
+# -*- encoding: utf8 -*-
+#
+# The Qubes OS Project, http://www.qubes-os.org
+#
+# Copyright (C) 2017 Marek Marczykowski-Górecki
+#                               <marmarek@invisiblethingslab.com>
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+import qubes.ext
+
+class CoreFeatures(qubes.ext.Extension):
+    # pylint: disable=too-few-public-methods
+    @qubes.ext.handler('features-request')
+    def qubes_features_request(self, vm, event, untrusted_features):
+        '''Handle features provided by qubes-core-agent and qubes-gui-agent'''
+        # pylint: disable=no-self-use,unused-argument
+        if getattr(vm, 'template', None):
+            vm.log.warning(
+                'Ignoring qubes.NotifyTools for template-based VM')
+            return
+
+        # for now used only to check for the tools presence
+        if 'version' in untrusted_features:
+            # any suspicious string will raise exception here,
+            # but otherwise ignored
+            int(untrusted_features['version'])
```

Comment:
There is nothing that prevents VM from exposing "1" or whatever it pleases without parsing this on dom0's end. When the need arises, someone will be rewriting the extension anyway, so we can just drop this now. Until then, we just ignore unknown keys, this one included.

As I said, I think it's not just useless, it actually sends unqualified "version" feature request to all extensions, without saying what exactly is versioned (gui? qrexec? the qubesdb handshake?).

This is because event handlers of features-request don't generally get context of the call and I'd imagine they contain something like:

```python
if 'my_expected_feature' in untrusted_features:
    # ...
```

From extension writer's perspective, there are some problems hiding around this handler. For one, we never promised that one `qubes.FeaturesRequest` will arrive as single event (and we obviously won't make such promise).

OTOH, other features' names (esp. "gui" and "qrexec") are good precisely because they are very general and I think more than one extension will have interest in them.